tempo_embeddings/embeddings/weaviate_database.py

```diff
@@ -583,3 +573,59 @@ def validate_config(self) -> None:
                     "Collection '%s' exists in the database but is not registered in the configuration database.",
                     collection,
                 )
+
+
+class QueryBuilder:
+    @staticmethod
+    def build_filter(
+        filter_words: list[str] = None,
+        year_from: Optional[int] = None,
+        year_to: Optional[int] = None,
+        metadata: Optional[dict[str, Any]] = None,
+        *,
+        text_field: str = "passage",
+        year_field: str = "year",
+    ) -> Optional[Filter]:
+        """Generic method to build a Weaviate Filter.
+
+        All filters are combined with AND.
+
+        Args:
+            filter_words (list[str], optional): Only include passage that contain any of these words. Defaults to None.
+            year_from (str, optional): Only include passages from this year or later. Defaults to None.
+            year_to (str, optional): Only include passages up to this year. Defaults to None.
+            metadata (dict[str, Any], optional): Additional filter criteria for exact matches in the form <field, term>.
+            text_field: The field name for the text. Defaults to "passage".
+            year_field (str, optional): The field name for the year. Defaults to "year".
+        Raises:
+            ValueError: If year_from > year_to
+        Returns:
+            Optional[Filter]: The filter object or None if none of the input arguments are set.
+
+        """
+        # TODO: contains_any should be a list of tuples to allow for multiple values per field
+
+        if year_from and year_to and year_from > year_to:
+            raise ValueError(
+                f"'year_from' ({year_from}) must be less than 'year_to' ({year_to})."
+            )
+
+        filters: list[Filter] = [
+            Filter.by_property(text_field).contains_any(word)
+            for word in (filter_words or [])
+        ]
+
+        if year_from is not None:
+            filters.append(Filter.by_property(year_field).greater_or_equal(year_from))
+        if year_to is not None:
+            filters.append(Filter.by_property(year_field).less_or_equal(year_to))
+
+        if metadata:
+            filters.extend(
+                [
+                    Filter.by_property(field).equal(value)
+                    for field, value in metadata.items()
+                ]
+            )
+
+        return Filter.all_of(filters) if filters else None
```

Comment:
Is this .all_of() like an AND? It would be nice to also have an filter builder with an OR version as well.

Reply:
Indeed, but I thought: let's expand this once we actually use it for something.